source/common/compressor/zlib_compressor_impl.h

```diff
@@ -61,28 +60,35 @@ class ZlibCompressorImpl : public Compressor {
             uint8_t memory_level);
 
   /**
-   * Finish must be called when the stream is over. It will compress all the remaining
-   * input data and flush it to the output buffer.
+   * Flush should be called when no more data needs to be compressed. It will compress
+   * any remaining input available in the compressor and flush the compressed data to the output
+   * buffer. Note that forcing flush frequently degrades the compression ratio, so this should only
+   * be called when necessary.
    * @param output_buffer supplies the buffer to output compressed data.
    */
-  void finish(Buffer::Instance& output_buffer);
+  void flush(Buffer::Instance& output_buffer);
+
+  /**
+   * @return uint64_t CRC-32 if a gzip stream is being written or Adler-32 for other compression
+   * types.
+   */
+  uint64_t checksum();
 
   /**
-   * Implements Envoy::Compressor
+   * Implements Envoy::Compressor.
    */
   void compress(const Buffer::Instance& input_buffer, Buffer::Instance& output_buffer) override;
 
 private:
-  void process(Buffer::Instance& output_buffer, uint8_t flush_state);
-  void commit(Buffer::Instance& output_buffer);
-  void reserve(Buffer::Instance& output_buffer);
+  bool deflateNext(int8_t flush_state);
+  void process(Buffer::Instance& output_buffer, int8_t flush_state);
+  void updateOutput(Buffer::Instance& output_buffer);
 
-  std::unique_ptr<Buffer::RawSlice> output_slice_ptr_;
-  std::unique_ptr<z_stream, std::function<void(z_stream*)>> zstream_ptr_;
-
-  bool initialized_{false};
+  uint64_t chunk_;
+  bool initialized_;
```

Comment:
is `initialized_` always `== output_char_ptr_ && zstream_ptr_` ?

Reply:
No, initialized_ is always false and then gets set to true after deflateInit2 is called. output_char_ptr_ && zstream_ptr_ are independent and cannot be used to tell whether or not zlib internal state has been initialized. That's the solution that I found for now, but I can look at that again.

Reply:
OK, given that I agree it makes sense to have the separate `initialized_` boolean.